private/buf/cmd/buf/command/push/push.go

```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+		)
+	}
+	return gitMetadataUploadOptions, nil
+}
+
+func getGitRemotes(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) ([]string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("remote"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return nil, err
+	}
+	return strings.Split(strings.TrimSpace(buffer.String()), "\n"), nil
+}
+
+// This returns an upload option for all Git metadata labels. We set labels for the following
+// Git matadata:
+//   - tags on current commit
+//   - current branch
+func getGitMetadataLabelsUploadOptions(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (bufmodule.UploadOption, error) {
+	tags, err := getGitTagsOnCurrentCommit(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	branch, err := getCurrentGitBranch(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	labels := append(tags, branch)
+	if len(labels) > 0 {
+		return bufmodule.UploadWithLabels(labels...), nil
+	}
+	return nil, nil
+}
+
+func getGitTagsOnCurrentCommit(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) ([]string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("tag", "--points-at", "HEAD"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return nil, err
+	}
+	if len(buffer.Bytes()) > 0 {
+		return strings.Split(strings.TrimSpace(buffer.String()), "\n"), nil
+	}
+	return nil, nil
+}
+
+func getCurrentGitBranch(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("branch", "--show-current"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(buffer.String()), nil
+}
+
+func getGitMetadataSourceControlURLUploadOption(
+	ctx context.Context,
+	runner command.Runner,
+	remotes []string,
+	input string,
+) (bufmodule.UploadOption, error) {
+	remoteToURL := map[string]string{}
+	for _, remote := range remotes {
+		buffer := bytes.NewBuffer(nil)
+		if err := runner.Run(
+			ctx,
+			gitCommand,
+			command.RunWithArgs("config", "--get", fmt.Sprintf("remote.%s.url", remote)),
+			command.RunWithStdout(buffer),
+			command.RunWithStderr(buffer),
+			command.RunWithDir(input),
+		); err != nil {
+			return nil, err
+		}
+		remoteToURL[remote] = strings.TrimSpace(buffer.String())
+	}
+	// We prioritize the Git default remote, "origin", URL
+	if gitDefaultRemoteURL, ok := remoteToURL[gitDefaultRemote]; ok {
+		// First we need to trim the `.git` suffix if there is one.
+		gitDefaultRemoteURL = strings.TrimSuffix(gitDefaultRemoteURL, ".git")
```

Comment:
I think we need to parse this URL. We will likely want to special-case based on hostname. This URL can be in a few different formats, see here: https://git-scm.com/docs/git-clone#_git_urls